Firestore/Swift/Source/PropertyWrapper/FirestoreQueryObservable.swift

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import SwiftUI
+import FirebaseFirestore
+
+@available(iOS 14.0, *)
+@available(tvOS, unavailable)
+internal class FirestoreQueryObservable<T>: ObservableObject {
+  @Published var items: T
+
+  private let firestore = Firestore.firestore()
+  private var listener: ListenerRegistration?
+
+  private var setupListener: (() -> Void)!
+
+  internal var configuration: FirestoreQuery<T>.Configuration {
+    didSet {
+      removeListener()
+      setupListener()
+    }
+  }
+
+  init<U: Decodable>(configuration: FirestoreQuery<T>.Configuration) where T == [U] {
+    items = []
+    self.configuration = configuration
+    setupListener = createListener { [weak self] querySnapshot, error in
+      if let _ = error {
+        // in this scenario, we actively ignore the error
+        self?.items = []
+        return
+      }
+
+      guard let documents = querySnapshot?.documents else {
+        self?.items = []
+        return
+      }
+
+      self?.items = documents.compactMap { document in
+        try? document.data(as: U.self)
+      }
+    }
+
+    setupListener()
+  }
+
+  init<U: Decodable>(configuration: FirestoreQuery<T>.Configuration) where T == [Result<U, Error>] {
+    items = []
+    self.configuration = configuration
+    setupListener = createListener { [weak self] querySnapshot, error in
+      if let error = error {
+        self?.items = [.failure(error)]
+        return
+      }
+
+      guard let documents = querySnapshot?.documents else {
+        self?.items = []
+        return
+      }
+
+      self?.items = documents.compactMap { queryDocumentSnapshot in
+        Result {
+          // NOTE: The `try` handles the parse failure
```

Comment:
I am not 100% sure this is correct. @schmidt-sebastian, can you comment?

Reply:
Documents in a QuerySnapshot always exist.

This is not true in a DocumentSnapshot (for single document fetches). We need to keep that in mind when expanding this to single document fetches.